File: src/corp.ts
/**
 * corp.ts — Cross-Origin-Resource-Policy per-sub-resource
 * audit. T76 cycle 27.
 *
 * CORP is the third member of the cross-origin-isolation triad
 * (with COOP + COEP). Where COOP controls window.opener and
 * COEP controls which sub-resources the page is willing to
 * embed, CORP is set on the RESOURCE side to opt INTO being
 * embedded by cross-origin pages.
 *
 * Threat model: a page with `Cross-Origin-Embedder-Policy:
 * require-corp` REQUIRES every cross-origin sub-resource it
 * fetches to carry a CORP header that allows the embedding
 * origin. Without CORP, the resource is BLOCKED by the
 * browser. So CORP-less sub-resources are:
 *
 *   * STRICT defects on a require-corp page (broken loads).
 *   * WARN defects on any page (forward-compat gap — the
 *     moment the page adopts COEP=require-corp, this resource
 *     stops working).
 *
 * CORP is also a Spectre-class side-channel mitigation in its
 * own right: when set to `same-origin`, browsers refuse to
 * even FETCH the resource cross-origin, which prevents some
 * cache-timing attacks that observe whether the resource was
 * already cached.
 *
 * Acceptable values (W3C):
 *   * `same-origin` — most restrictive; resource only loadable
 *                     by same-origin pages.
 *   * `same-site`   — same eTLD+1.
 *   * `cross-origin` — anyone can embed (CDN-style).
 *
 * Findings:
 *
 *   - corp.cross-origin-resource-no-corp     (warn or strict)
 *     A cross-origin sub-resource the page fetched lacks a
 *     CORP header. Strict if the page itself sets COEP=
 *     require-corp (the resource will be blocked at load);
 *     warn otherwise.
 *
 *   - corp.cross-origin-resource-invalid     warn
 *     A cross-origin sub-resource has CORP set to a value
 *     not in the W3C-recognised set. Browsers may reject the
 *     resource entirely.
 *
 * Out of scope:
 *   * Same-origin sub-resources — CORP doesn't apply.
 *   * The page's OWN top-level CORP — separate concern (most
 *     top-level HTML pages legitimately don't set CORP). A
 *     future top-level CORP detector could surface
 *     `CORP: cross-origin` on a top-level HTML page where
 *     `same-origin` would be safer; not in this cycle.
 *   * Localhost (consistent with the response-header detector
 *     family).
 *   * data: / blob: / about: URLs (no transport).
 *
 * Architecture: this is the FIRST per-sub-resource detector.
 * It reads from a NEW capture path — `allResponseHeaders` Map
 * — populated by the response listener in main.ts for every
 * response (not just top-level navigation). The existing
 * response-header detectors keep using `topLevelResponseHeaders`.
 *
 * The CORP detector takes BOTH:
 *   * `pageUrl` — to determine same-origin vs cross-origin.
 *   * `pageCoepValue` — the page's own COEP header value, to
 *     decide whether missing CORP is strict or warn.
 *   * `allResponseHeaders` — Map of every sub-resource's
 *     headers.
 *
 * It does NOT use the existing `responseHeaderDetector` helper
 * because the helper's contract is "one snapshot per page
 * navigation, classify into findings". CORP's contract is
 * "walk every sub-resource, classify each, aggregate". A new
 * shape — kept bespoke for now per the cycle-22 verdict on
 * heterogeneous classifier shapes.
 */

export interface CorpFinding {
  severity: 'strict' | 'warn';
  kind: string;
  detail: string;
  evidence: Record<string, unknown>;
}

export interface CapturedSubResource {
  url: string;
  origin: string;
  /** Lowercased + trimmed CORP value, or null if absent. */
  corp: string | null;
}

export interface CorpSnapshot {
  pageUrl: string;
  pageOrigin: string;
  pageIsLocalhost: boolean;
  /** True iff the page's own COEP is 'require-corp'. */
  pageRequiresCorp: boolean;
  /** Cross-origin sub-resources only — same-origin filtered out. */
  crossOriginSubResources: CapturedSubResource[];
}

const ACCEPTABLE_CORP: ReadonlySet<string> = new Set([
  'same-origin',
  'same-site',
  'cross-origin',
]);

function isLocalhost(url: string): boolean {
  try {
    const u = new URL(url);
    const h = u.hostname;
    return h === 'localhost' || h === '127.0.0.1' || h === '::1' || h.endsWith('.localhost');
  } catch {
    return false;
  }
}

function originOf(url: string): string {
  try {
    const u = new URL(url);
    return `${u.protocol}//${u.host}`;
  } catch {
    return '';
  }
}

function isAuditableScheme(url: string): boolean {
  return url.startsWith('http://') || url.startsWith('https://');
}

function getHeaderValue(
  headers: Record<string, string> | undefined,
  name: string,
): string | null {
  if (!headers) return null;
  for (const [k, v] of Object.entries(headers)) {
    if (k.toLowerCase() === name) return v;
  }
  return null;
}

export function buildCorpSnapshot(
  pageUrl: string,
  pageHeaders: Record<string, string> | undefined,
  allHeaders: Map<string, Record<string, string>>,
): CorpSnapshot {
  const pageOrigin = originOf(pageUrl);
  const pageIsLocalhost = isLocalhost(pageUrl);
  const coep = getHeaderValue(pageHeaders, 'cross-origin-embedder-policy');
  const pageRequiresCorp = coep !== null && coep.trim().toLowerCase() === 'require-corp';

  const crossOrigin: CapturedSubResource[] = [];
  for (const [url, headers] of allHeaders.entries()) {
    if (!isAuditableScheme(url)) continue;
    const ro = originOf(url);
    if (!ro || ro === pageOrigin) continue;
    const corp = getHeaderValue(headers, 'cross-origin-resource-policy');
    crossOrigin.push({
      url,
      origin: ro,
      corp: corp === null ? null : corp.trim().toLowerCase(),
    });
  }

  return {
    pageUrl,
    pageOrigin,
    pageIsLocalhost,
    pageRequiresCorp,
    crossOriginSubResources: crossOrigin,
  };
}

export function detectCorpIssues(snap: CorpSnapshot): CorpFinding[] {
  if (snap.pageIsLocalhost) return [];

  const noCorp: CapturedSubResource[] = [];
  const invalid: CapturedSubResource[] = [];

  for (const r of snap.crossOriginSubResources) {
    if (r.corp === null) {
      noCorp.push(r);
    } else if (!ACCEPTABLE_CORP.has(r.corp)) {
      invalid.push(r);
    }
  }

  const out: CorpFinding[] = [];
  const renderEx = (r: CapturedSubResource) => `${r.origin} ← '${r.url}'`;

  if (noCorp.length > 0) {
    const examples = noCorp.slice(0, 5).map(renderEx);
    const severity: 'strict' | 'warn' = snap.pageRequiresCorp ? 'strict' : 'warn';
    const blocker = snap.pageRequiresCorp
      ? 'The page sets Cross-Origin-Embedder-Policy: require-corp, so the browser BLOCKS these resources at load — they fail to render.'
      : 'The page does not currently enforce COEP=require-corp, so the resources still load. The moment the page adopts COEP=require-corp (a supersociety baseline for any app handling sensitive data), every one of these sub-resources stops working.';
    out.push({
      severity,
      kind: 'corp.cross-origin-resource-no-corp',
      detail: `${noCorp.length} cross-origin sub-resource(s) lack a Cross-Origin-Resource-Policy header. ${blocker} The fix lives on the SERVER side of each resource — set 'Cross-Origin-Resource-Policy: cross-origin' on the asset response (or 'same-site' / 'same-origin' for tighter scoping). Examples: ${examples.join('; ')}`,
      evidence: {
        count: noCorp.length,
        examples,
        pageRequiresCorp: snap.pageRequiresCorp,
      },
    });
  }

  if (invalid.length > 0) {
    const examples = invalid.slice(0, 5).map((r) => `${renderEx(r)} (CORP='${r.corp}')`);
    out.push({
      severity: 'warn',
      kind: 'corp.cross-origin-resource-invalid',
      detail: `${invalid.length} cross-origin sub-resource(s) have a Cross-Origin-Resource-Policy header set to a value not in the W3C-recognised set ('same-origin', 'same-site', 'cross-origin'). Browsers may reject the resource entirely. Examples: ${examples.join('; ')}`,
      evidence: { count: invalid.length, examples },
    });
  }

  return out;
}

File: crates/crawler-detectors/src/corp.rs
//! `corp` — per-sub-resource Cross-Origin-Resource-Policy audit.
//!
//! Mirror of `src/corp.ts`. First per-sub-resource detector (vs the
//! per-page-headers pattern the other ports use). The crawler-side
//! `allHeaders` Map carries every sub-resource response's headers;
//! this detector walks them, filters to cross-origin loads, and
//! classifies each.
//!
//! Findings:
//!
//!   * `corp.cross-origin-resource-no-corp` — strict if the page
//!     declares COEP=require-corp (browser BLOCKS), warn otherwise
//!     (forward-compat: the resource breaks the moment the page
//!     adopts require-corp)
//!   * `corp.cross-origin-resource-invalid` — CORP value not in
//!     {same-origin, same-site, cross-origin}; browsers may reject
//!
//! AVP-2 invariants: `unsafe_code = "deny"`, pure detector, no I/O.

use crate::url_helpers::is_localhost;
use crate::{AxisFinding, AxisSeverity};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

/// One captured cross-origin sub-resource the page loaded.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct CapturedSubResource {
    /// Full URL of the sub-resource.
    pub url: String,
    /// `<scheme>://<host>` of the sub-resource (cached for grouping).
    pub origin: String,
    /// Lowercase + trimmed CORP value, or `None` if absent.
    pub corp: Option<String>,
}

/// Captured page state the detector consumes.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
#[non_exhaustive]
pub struct CorpSnapshot {
    /// Top-level page URL.
    pub page_url: String,
    /// `<scheme>://<host>` of the page.
    pub page_origin: String,
    /// Localhost / loopback exemption.
    pub page_is_localhost: bool,
    /// True iff the page's COEP header is exactly `require-corp`.
    pub page_requires_corp: bool,
    /// Cross-origin sub-resources only — same-origin filtered out
    /// at snapshot-build time.
    pub cross_origin_sub_resources: Vec<CapturedSubResource>,
}

/// Acceptable CORP values per W3C.
const ACCEPTABLE_CORP: &[&str] = &["same-origin", "same-site", "cross-origin"];

/// Extract `<scheme>://<host>` from a URL string. Returns empty
/// string on parse failure.
fn origin_of(url: &str) -> String {
    let Some(scheme_end) = url.find("://") else {
        return String::new();
    };
    let scheme = &url[..scheme_end];
    let after = &url[scheme_end + 3..];
    let host_end = after
        .find('/')
        .or_else(|| after.find('?'))
        .or_else(|| after.find('#'))
        .unwrap_or(after.len());
    format!("{}://{}", scheme, &after[..host_end])
}

/// Only http(s) sub-resources are auditable; data:/blob:/about:
/// have no transport-level CORP.
fn is_auditable_scheme(url: &str) -> bool {
    url.starts_with("http://") || url.starts_with("https://")
}

/// Build a snapshot from the captured per-resource headers map.
/// `all_headers` maps URL → that response's lowercased header map.
pub fn build_corp_snapshot(
    page_url: &str,
    page_headers: &BTreeMap<String, String>,
    all_headers: &BTreeMap<String, BTreeMap<String, String>>,
) -> CorpSnapshot {
    let page_origin = origin_of(page_url);
    let page_is_localhost = is_localhost(page_url);

    let coep = page_headers
        .iter()
        .find(|(k, _)| k.eq_ignore_ascii_case("cross-origin-embedder-policy"))
        .map(|(_, v)| v.trim().to_ascii_lowercase());
    let page_requires_corp = coep.as_deref() == Some("require-corp");

    let mut cross_origin: Vec<CapturedSubResource> = Vec::new();
    for (url, headers) in all_headers {
        if !is_auditable_scheme(url) {
            continue;
        }
        let ro = origin_of(url);
        if ro.is_empty() || ro == page_origin {
            continue;
        }
        let corp = headers
            .iter()
            .find(|(k, _)| k.eq_ignore_ascii_case("cross-origin-resource-policy"))
            .map(|(_, v)| v.trim().to_ascii_lowercase());
        cross_origin.push(CapturedSubResource {
            url: url.clone(),
            origin: ro,
            corp,
        });
    }

    CorpSnapshot {
        page_url: page_url.to_owned(),
        page_origin,
        page_is_localhost,
        page_requires_corp,
        cross_origin_sub_resources: cross_origin,
    }
}

/// Render an example string for one sub-resource finding.
fn render_example(r: &CapturedSubResource) -> String {
    format!("{} ← '{}'", r.origin, r.url)
}

/// Pure detector: snapshot → findings. No I/O.
pub fn detect_corp_issues(snap: &CorpSnapshot) -> Vec<AxisFinding> {
    if snap.page_is_localhost {
        return Vec::new();
    }

    let mut no_corp: Vec<&CapturedSubResource> = Vec::new();
    let mut invalid: Vec<&CapturedSubResource> = Vec::new();

    for r in &snap.cross_origin_sub_resources {
        match r.corp.as_deref() {
            None => no_corp.push(r),
            Some(v) if !ACCEPTABLE_CORP.contains(&v) => invalid.push(r),
            Some(_) => {} // valid value → no finding
        }
    }

    let mut out = Vec::new();

    if !no_corp.is_empty() {
        let examples: Vec<String> = no_corp.iter().take(5).map(|r| render_example(r)).collect();
        let severity = if snap.page_requires_corp {
            AxisSeverity::Strict
        } else {
            AxisSeverity::Warn
        };
        let blocker = if snap.page_requires_corp {
            "The page sets Cross-Origin-Embedder-Policy: require-corp, so the browser BLOCKS these resources at load — they fail to render."
        } else {
            "The page does not currently enforce COEP=require-corp, so the resources still load. The moment the page adopts COEP=require-corp (a supersociety baseline for any app handling sensitive data), every one of these sub-resources stops working."
        };
        out.push(AxisFinding {
            severity,
            kind: "corp.cross-origin-resource-no-corp".into(),
            detail: format!(
                "{} cross-origin sub-resource(s) lack a Cross-Origin-Resource-Policy header. {} The fix lives on the SERVER side of each resource — set 'Cross-Origin-Resource-Policy: cross-origin' on the asset response (or 'same-site' / 'same-origin' for tighter scoping). Examples: {}",
                no_corp.len(),
                blocker,
                examples.join("; ")
            ),
        });
    }

    if !invalid.is_empty() {
        let examples: Vec<String> = invalid
            .iter()
            .take(5)
            .map(|r| {
                format!(
                    "{} (CORP='{}')",
                    render_example(r),
                    r.corp.as_deref().unwrap_or("")
                )
            })
            .collect();
        out.push(AxisFinding {
            severity: AxisSeverity::Warn,
            kind: "corp.cross-origin-resource-invalid".into(),
            detail: format!(
                "{} cross-origin sub-resource(s) have a Cross-Origin-Resource-Policy header set to a value not in the W3C-recognised set ('same-origin', 'same-site', 'cross-origin'). Browsers may reject the resource entirely. Examples: {}",
                invalid.len(),
                examples.join("; ")
            ),
        });
    }

    out
}

#[cfg(test)]
mod tests {
    use super::*;

    fn page_headers(require_corp: bool) -> BTreeMap<String, String> {
        let mut m = BTreeMap::new();
        if require_corp {
            m.insert("cross-origin-embedder-policy".into(), "require-corp".into());
        }
        m
    }

    fn build(
        page_url: &str,
        require_corp: bool,
        sub_resources: &[(&str, Option<&str>)],
    ) -> CorpSnapshot {
        let mut all = BTreeMap::new();
        for (url, corp) in sub_resources {
            let mut h = BTreeMap::new();
            if let Some(v) = corp {
                h.insert("cross-origin-resource-policy".into(), (*v).to_owned());
            }
            all.insert((*url).to_owned(), h);
        }
        build_corp_snapshot(page_url, &page_headers(require_corp), &all)
    }

    #[test]
    fn localhost_skipped() {
        let s = build("http://localhost/", false, &[]);
        assert!(detect_corp_issues(&s).is_empty());
    }

    #[test]
    fn same_origin_filtered_out_at_snapshot_time() {
        let s = build(
            "https://example.com/",
            false,
            &[("https://example.com/img.png", None)],
        );
        assert_eq!(s.cross_origin_sub_resources.len(), 0);
        assert!(detect_corp_issues(&s).is_empty());
    }

    #[test]
    fn cross_origin_no_corp_warns_without_require_corp() {
        let s = build(
            "https://example.com/",
            false,
            &[("https://cdn.example.org/img.png", None)],
        );
        let f = detect_corp_issues(&s);
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].kind, "corp.cross-origin-resource-no-corp");
        assert_eq!(f[0].severity, AxisSeverity::Warn);
    }

    #[test]
    fn cross_origin_no_corp_strict_when_page_requires_corp() {
        let s = build(
            "https://example.com/",
            true,
            &[("https://cdn.example.org/img.png", None)],
        );
        let f = detect_corp_issues(&s);
        assert_eq!(f.len(), 1);
        assert_eq!(f[0].severity, AxisSeverity::Strict);
        assert!(f[0].detail.contains("BLOCKS"));
    }

    #[test]
    fn acceptable_corp_values_pass() {
        for val in ["same-origin", "same-site", "cross-origin"] {
            let s = build(
                "https://example.com/",
                false,
                &[("https://cdn.example.org/img.png", Some(val))],
            );
            assert!(
                detect_corp_issues(&s).is_empty(),
                "{val} should be acceptable"
            );
        }
    }

    #[test]
    fn corp_case_insensitive() {
        let s = build(
            "https://example.com/",
            false,
            &[("https://cdn.example.org/img.png", Some("SAME-ORIGIN"))],
        );
        assert!(detect_corp_issues(&s).is_empty());
    }

    #[test]
    fn invalid_corp_value_warns() {
        let s = build(
            "https://example.com/",
            false,
            &[("https://cdn.example.org/img.png", Some("anywhere"))],
        );
        let f = detect_corp_issues(&s);
        assert!(f
            .iter()
            .any(|x| x.kind == "corp.cross-origin-resource-invalid"));
    }

    #[test]
    fn data_url_subresource_skipped() {
        let s = build(
            "https://example.com/",
            false,
            &[("data:image/png;base64,AAAA", None)],
        );
        assert!(detect_corp_issues(&s).is_empty());
    }

    #[test]
    fn multiple_subresources_aggregate_in_one_finding() {
        let s = build(
            "https://example.com/",
            false,
            &[
                ("https://cdn.a.org/1.png", None),
                ("https://cdn.b.org/2.png", None),
                ("https://cdn.c.org/3.png", None),
            ],
        );
        let f = detect_corp_issues(&s);
        assert_eq!(f.len(), 1);
        assert!(f[0].detail.contains("3 cross-origin"));
    }

    #[test]
    fn examples_capped_at_5() {
        let urls: Vec<(String, Option<&str>)> = (0..10)
            .map(|i| (format!("https://cdn.{i}.org/x.png"), None))
            .collect();
        let url_refs: Vec<(&str, Option<&str>)> =
            urls.iter().map(|(u, c)| (u.as_str(), *c)).collect();
        let s = build("https://example.com/", false, &url_refs);
        let f = detect_corp_issues(&s);
        assert!(f[0].detail.contains("10 cross-origin"));
        // 5 examples in the rendered list
        let count = f[0].detail.matches("← '").count();
        assert_eq!(count, 5);
    }
}
